lib/prometheus-parser/src/line.rs

```diff
@@ -0,0 +1,674 @@
+                        Ok((inner_input, _)) => inner_input,
+                        Err(sep_err) => match match_char('}')(inner_input) {
+                            Ok((inner_input, _)) => {
+                                input = inner_input;
+                                break;
+                            }
+                            Err(_) => return Err(sep_err),
+                        },
+                    };
+
+                    input = inner_input;
+                }
+            }
+        }
+        Ok((input, result))
+    }
+
+    /// Parse `{label_name="value",...}`
+    fn parse_labels(input: &str) -> IResult<BTreeMap<String, String>> {
+        let input = trim_space(input);
+
+        match opt(char('{'))(input) {
+            Ok((input, None)) => Ok((input, BTreeMap::new())),
+            Ok((input, Some(_))) => Self::parse_labels_inner(input),
+            Err(failure) => Err(failure),
+        }
+    }
+
+    /// Parse `'"' string_content '"'`. `string_content` can contain any unicode characters,
+    /// backslash (`\`), double-quote (`"`), and line feed (`\n`) characters have to be
+    /// escaped as `\\`, `\"`, and `\n`, respectively.
+    fn parse_escaped_string(input: &str) -> IResult<String> {
+        #[derive(Debug)]
+        enum StringFragment<'a> {
+            Literal(&'a str),
+            EscapedChar(char),
+        }
+
+        let parse_string_fragment = alt((
+            map(is_not("\"\\"), StringFragment::Literal),
+            map(
+                preceded(
+                    char('\\'),
+                    alt((
+                        value('\n', char('n')),
+                        value('"', char('"')),
+                        value('\\', char('\\')),
+                    )),
+                ),
+                StringFragment::EscapedChar,
+            ),
+        ));
+
+        let input = trim_space(input);
+
+        let build_string = fold_many0(
+            parse_string_fragment,
+            String::new(),
+            |mut result, fragment| {
+                match fragment {
+                    StringFragment::Literal(s) => result.push_str(s),
+                    StringFragment::EscapedChar(c) => result.push(c),
+                }
+                result
+            },
+        );
+
+        fn match_quote(input: &str) -> IResult<char> {
+            char('"')(input).map_err(|_: NomError| {
+                ErrorKind::ExpectedToken {
+                    expected: "\"",
+                    input: input.to_owned(),
+                }
+                .into()
+            })
+        }
+
+        delimited(match_quote, build_string, match_quote)(input)
+    }
+}
+
+impl Header {
+    fn space1(input: &str) -> IResult<()> {
+        take_while1(|c| c == ' ' || c == '\t')(input)
+            .map_err(|_: NomError| {
+                ErrorKind::ExpectedSpace {
+                    input: input.to_owned(),
+                }
+                .into()
+            })
```

Comment:
Nom errors are single enum variants (https://docs.rs/nom/5.1.2/nom/error/enum.ErrorKind.html), it is not very meaningful and I don't think we should include it. In this case it is an `ErrorKind::Takewhile1`.

Reply:
Ok, got it. FTR @Jeffail went with [pest](https://crates.io/crates/pest) in part because of the error message story. No need for you to convert though, it would probably be overkill here.